p4-fusion/p4_api.cc

```diff
@@ -191,48 +185,58 @@ ChangesResult P4API::Changes(const std::string& path, const std::string& from, i
 
 	args.push_back(path + pathAddition);
 
-	ChangesResult result = Run<ChangesResult>("changes", args);
+	ChangesResult result = Run<ChangesResult>("changes", args, []() -> ChangesResult
+	    { return {}; });
 
 	return result;
 }
 
-DescribeResult P4API::Describe(const std::string& cl)
+DescribeResult P4API::Describe(GitAPI& git, const int cl)
 {
 	MTR_SCOPE("P4", __func__);
+
 	return Run<DescribeResult>("describe", { "-s", // Omit the diffs
-	                                           cl });
+	                                           std::to_string(cl) },
+	    [&git]() -> DescribeResult
+	    { return { git }; });
 }
 
-FileLogResult P4API::FileLog(const std::string& changelist)
+FileLogResult P4API::FileLog(GitAPI& git, const int changelist)
 {
 	return Run<FileLogResult>("filelog", {
 	                                         "-c", // restrict output to a single changelist
-	                                         changelist,
+	                                         std::to_string(changelist),
 	                                         "-m1", // don't get the full history, just the first entry.
 	                                         "//..." // rather than require the path to be passed in, just list all files.
-	                                     });
+	                                     },
+	    [&git]() -> FileLogResult
+	    { return { git }; });
 }
 
-PrintResult P4API::PrintFiles(const std::vector<std::string>& fileRevisions)
+PrintResult P4API::PrintFiles(const std::vector<std::string>& fileRevisions, const std::function<void()>& onStat, const std::function<void(const char*, int)>& onOutput)
 {
 	MTR_SCOPE("P4", __func__);
 
 	if (fileRevisions.empty())
 	{
-		return PrintResult();
+		return { onStat, onOutput };
```

Comment:
This function is only used from one place, so this code path will never be hit because there is a fast path earlier for `fileRevisions.empty()` before this is called. It would be better to instead document that as a pre-condition on this function and add an `assert(!fileRevisions.empty() && "expected non-empty list of file revisions");`